source/slang/parser.cpp

```diff
@@ -1004,29 +1099,62 @@ namespace Slang
         }
     }
 
-    static void ParseFuncDeclHeader(
+    static RefPtr<Decl> ParseFuncDeclHeader(
         Parser*                     parser,
         DeclaratorInfo const&       declaratorInfo,
-        RefPtr<FuncDecl>  decl)
+        RefPtr<FuncDecl>  decl,
+        RefPtr<GenericDecl> genDecl)
     {
-        parser->PushScope(decl.Ptr());
+        RefPtr<Decl> retDecl = decl;
 
         parser->FillPosition(decl.Ptr());
         decl->loc = declaratorInfo.nameAndLoc.loc;
 
         decl->nameAndLoc = declaratorInfo.nameAndLoc;
+
+        // if return type is a DeclRef type, we need to update its scope to use this function decl's scope
+        // so that LookUp can find the generic type parameters declared after the function name
+        if (auto declRefRetType = declaratorInfo.typeSpec.As<DeclRefExpr>())
+            declRefRetType->scope = parser->currentScope;
+
         decl->ReturnType = TypeExp(declaratorInfo.typeSpec);
-        parseParameterList(parser, decl);
-        ParseOptSemantics(parser, decl.Ptr());
+        auto parseFuncDeclHeaderInner = [&](GenericDecl *)
+        {
+            parseParameterList(parser, decl);
+            ParseOptSemantics(parser, decl.Ptr());
+            return decl;
+        };
+
+        if (parser->LookAheadToken(TokenType::OpLess))
```

Comment:
Minor point: checking for `OpLess` specifically here means that we need to be careful when defining generic operator overloads:

    bool operator<<T>(Foo<T> left, Foo<T> right);

In the above example, I am trying to define a generic (`<T>`) operator (`operator<`) but the parser is going to get confused unless I add spaces.

Long (long) term, a more robust check here is to see if *either* the next token is a `<` (or maybe even just an operator that *begins* with `<`), *or* the function name is an `operator` name where the operator *ends* with `<` (which we'd trim from the name). Of course that then creates an ambiguity when the user really wanted to define a non-generic `operator<<`, so we'd need to have some disambiguation logic for that case (similar to disambiguation at a generic use site).